server/src/test/java/org/elasticsearch/index/mapper/DynamicTemplatesTests.java

```diff
@@ -2252,4 +2373,137 @@ public void testInvalidMatchWithArrayOfFieldNamesUsingNonStringEntries() throws
             containsString("[match] values must either be a string or list of strings, but was [[23.45, false]]")
         );
     }
+
+    public void testIsValidPatternForDefaultMatchType() {
+        record MatchPattern(String pattern, boolean isValid) {}
+
+        MatchPattern[] candidates = new MatchPattern[] {
+            // valid match_pattern=DEFAULT entries
+            new MatchPattern("", true),    // has no metacharacters
+            new MatchPattern("foo", true), // has no metacharacters
+            new MatchPattern("user.name", true), // regular ES dotted field, so not a "field regex"
+            new MatchPattern("*foo", true),
+            new MatchPattern("foo*", true),       // matches simple wildcard, so not considered an ES field regex
+            new MatchPattern("foo?", true),
+            new MatchPattern("foo+", true),
+            new MatchPattern("a(cb)", true),
+            new MatchPattern("*.*", true),        // legitimate simple pattern for dotted path name
+            new MatchPattern("foo.*.bar", true),  // legitimate simple pattern for dotted path name
+            new MatchPattern(".middle*", true),
+            new MatchPattern("zero.one*", true),
+            new MatchPattern("*two.three$", true), // doesn't compile to regex
+            new MatchPattern("$^[foo", true),      // doesn't compile to regex
+
+            // invalid match_pattern=DEFAULT entries
+            new MatchPattern(".*foo", false),
+            new MatchPattern("^foo", false),
+            new MatchPattern("foo$", false),
+            new MatchPattern("f.*oo$", false),
+            new MatchPattern("a|b", false),
+            new MatchPattern("a[cb]", false),
+            new MatchPattern("a{1,2}", false),
+            new MatchPattern("[xyz]*.*", false) };
+
+        DynamicTemplate.MatchType matchType = DynamicTemplate.MatchType.DEFAULT;
+
+        for (MatchPattern c : candidates) {
+            if (c.isValid) {
+                // should not throw an Exception
+                matchType.isValidPattern(c.pattern);
+            } else {
+                Exception e = expectThrows(
+                    IllegalArgumentException.class,
+                    "pattern tested: " + c.pattern,
+                    () -> matchType.isValidPattern(c.pattern)
+                );
+                assertThat(
+                    e.getMessage(),
+                    containsString("Pattern [" + c.pattern + "] appears to be a regular expression, not a simple wildcard pattern")
+                );
+            }
+        }
+    }
+
+    public void testIsValidPatternForSimpleMatchType() {
+        record MatchPattern(String pattern) {}
```

Comment:
I'm not sure this adds anything over just using `String` here?

Reply:
Right. I'll change that.